src/lib/secret.ts

```diff
@@ -0,0 +1,83 @@
+/**
+ * Secrets can be in one of two modes:
+ *
+ * 1. A secret from ~/.fauna-shell. These secrets have a fauna key, like
+ * `fn1234` in the `key` field, and `allowScope` is true. Scopes can be appended
+ * to these keys.
+ *
+ * 2. A secret from `--secret`. Because the user is explicitly providing a
+ * secret, we do not attempt to parse or verify this secret. So the entire
+ * secret, including a database path, is stored in `key` (like
+ * `fn1234:db:@role/foo`). This type of secret has `allowScope` set to `false`,
+ * and appending a scope will throw an error.
+ */
+export class Secret {
+  // A fauna key, like `fn1234`.
+  key: string;
+  // A database scope, like `["foo", "bar"]`
+  databaseScope: string[];
+  // If we passed a secret with `--secret`, we disallow setting a scope or role.
+  allowScope: boolean;
+
+  constructor(opts: {
+    key: string;
+    allowScope: boolean;
+    databaseScope?: string[];
+  }) {
+    this.key = opts.key;
+    this.allowScope = opts.allowScope;
+    this.databaseScope = opts.databaseScope ?? [];
+  }
+
+  static fromFlag(key: string) {
+    if (key.length === 0) {
+      throw new Error("Secret cannot be empty");
+    }
+    return new Secret({ key, allowScope: !key.includes(":") });
+  }
+
+  static parse(key: string) {
```

Comment:
does the 'current' shell allow you to have endpoints in your file that are scoped?
```
[endpoint.child]
secret=secret:child:admin
secret=secret:child:@role/myrole
```

maybe we don't care? just trying to think through if this could lead to backwards incompatible changes for users and where we draw the line on that.

Reply:
the current shell lets you do that, but it will break if you set a database path in `fauna eval` or `fauna shell`. It seems like something we ought to just disallow.